src/transforms/sample/transform.rs

```diff
@@ -12,24 +12,86 @@ use crate::{
 use vector_lib::lookup::lookup_v2::OptionalValuePath;
 use vector_lib::lookup::OwnedTargetPath;
 
+/// Exists only for backwards compatability purposes so that the value of sample_rate_key is
+/// consistent after the internal implementation of the Sample class was modified to work in terms
+/// of percentages
+#[derive(Clone, Debug)]
+pub enum SampleMode {
+    Rate(u64, HashMap<Option<String>, u64>),
+    Ratio(f64, HashMap<Option<String>, f64>),
+}
+
+impl SampleMode {
+    pub fn new_rate(rate: u64) -> Self {
+        Self::Rate(rate, HashMap::default())
+    }
+
+    pub fn new_ratio(ratio: f64) -> Self {
+        Self::Ratio(ratio, HashMap::default())
+    }
+
+    fn increment(&mut self, key: &Option<String>) -> bool {
+        match self {
+            Self::Rate(rate, counter) => {
+                let counter_value = counter.entry(key.clone()).or_default();
+                let increment: u64 = *counter_value + 1;
+                *counter_value = increment;
+                increment % *rate == 0
+            }
+            Self::Ratio(ratio, counter) => {
+                let counter_value = counter.entry(key.clone()).or_insert(*ratio);
+                let increment: f64 = *counter_value + *ratio;
+                *counter_value = if increment >= 1.0 {
+                    increment - 1.0
+                } else {
+                    increment
+                };
+                increment >= 1.0
+            }
+        }
+    }
+
+    fn hash_within_ratio(&self, value: &[u8]) -> bool {
+        let hash = seahash::hash(value);
+        match self {
+            Self::Rate(rate, _) => hash % rate == 0,
+            Self::Ratio(ratio, _) => {
+                // Assuming an even distribution of values, process the event if the value of its hash %
+                // 100, is within the allowable configured ratio
+                (hash % 100) as f64 <= (ratio * 100.0)
```

Comment:
This calculation cannot produce correct values for ratios with more than two digits of precision. Mathematically, what you want is: `(hash as f64) / (u64::MAX as f64) <= ratio`, but that's not right either because `u64::MAX as f64` truncates some of the lower bits of `u64::MAX` (all-ones). Instead, this needs: `(hash as f64) / ((u64::MAX as u128 + 1) as f64) <= ratio` since that value is just a single one bit which does not lose precision. Moving the constants all to one side, you get: `hash <= (ratio * ((u64::MAX as u128) + 1) as f64) as u64`. That's pretty fugly, so you may want to pre-compute that constant and store it in the struct as well.

Reply:
Nice catch